### ISSUE TITLE
alse positive `invalid-argument-type` ("does not satisfy upper bound ... of type variable `Self`") when a generic class bounded by `NamedTuple` constructs itself

### ISSUE THREAD

**Issue by nverchev** [astral-sh/ruff#28589]
### Summary

**Code:**

```python
from collections.abc import Callable
from typing import Generic, NamedTuple, TypeVar

T = TypeVar('T', bound=NamedTuple)
U = TypeVar('U', bound=list[int])


class Box(Generic[T]):
    def __init__(self, *fn: Callable[[T], int]) -> None:
        self.fn = tuple(fn)

    def combine(self, other: 'Box[T]') -> 'Box[T]':
        return Box(*self.fn, *other.fn)  # error


class ListBox(Generic[U]):
    def __init__(self, *fn: Callable[[U], int]) -> None:
        self.fn = tuple(fn)

    def combine(self, other: 'ListBox[U]') -> 'ListBox[U]':
        return ListBox(*self.fn, *other.fn)  # no error
```

**Actual:**

```
error[invalid-argument-type]: Argument to `Box.__init__` is incorrect
Argument type `Box[T@Box]` does not satisfy upper bound `Box[T@Box]` of type variable `Self`
```

**Expected:** no error, as for `ListBox`, which differs only in the bound. The same happens with `bound=Union[int, NamedTuple]`. Pyright (basedpyright, standard mode) reports no error for either class.

**Related:** possibly the same root cause as #3907 (class-scoped generic bounded by `NamedTuple`), which was closed as fixed by astral-sh/ruff#26701. That PR was merged in July 2026 and is included in 0.0.80, but this case still fails there; it involves no protocol.

### Version

0.0.80
